keps/sig-api-machinery/1965-kube-apiserver-identity/README.md

```diff
@@ -95,20 +120,21 @@ will only delay the storage migration for the same period of time.
 
 ## Design Details
 
-The [kubelet heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md)
+The [kubelet heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/589-efficient-node-heartbeats)
 logic [already written](https://github.com/kubernetes/kubernetes/tree/master/pkg/kubelet/nodelease)
 will be re-used. The heartbeat controller will be added to kube-apiserver in a
 post-start hook.
 
-Each kube-apiserver will refresh its Lease every 10s by default. A GC controller
-will watch the Lease API using an informer, and periodically resync its local
-cache. On processing an item, the controller will delete the Lease if the last
-`renewTime` was more than `leaseDurationSeconds` ago (default to 1h). The
-default `leaseDurationSeconds` is chosen to be way longer than the default
+Each kube-apiserver will run a lease controller in a post-start-hook to refresh
```

Comment:
Leases have a thing called graceful release in the standard client-go code.  For beta, can we leverage this mechanism so that a server in a controlled shutdown can remove itself?  Do we not want graceful release for cases where we expect to be restarted and desire our indicated state of the world to persist over the gap?

Reply:
I think I've settled on wanting the apiserver identity to persist over the gap because we'd generally expect consistency of the identity on a host or member over time.

@lavalamp make sense to you too?

Reply:
> I think I've settled on wanting the apiserver identity to persist over the gap because we'd generally expect consistency of the identity on a host or member over time.

So to clarify, we're saying the holder identity should not change over a restart? i.e. remove the PID from the ID?

Reply:
Turns out the actual ID being used is just `kube-apiserver-<uuid>` (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config.go#L326). We'll have to revisit that during implementation.

Reply:
If we want identity to persist then we'd need to make a bunch of changes like not put the PID in the identity -- we put that in there in case you run two apiservers on the same system.

I don't think I want people to get attached to specific apiservers (don't make them pets) but it is handy to have a more durable name so you can find the thing or notice if one is consistently misbehaving. I'd personally keep this lease ephemeral but consider adding a nickname to help with readability / persistence over time.

Reply:
Maybe a `kubernetes.io/hostname` label on the Lease might be sufficient?

I see value in having the lease naming mirror what already exists for kube-controller-manager and kube-scheduler for consistency.